docs/rfcs/0006-ada-dynamic.md

```diff
@@ -0,0 +1,262 @@
+
+### Examples
+This section shows some examples for each strategies.
+
+**The `THRESHOLD` strategy:**
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  pipeline:
+    stages:
+      - name: K8S_PRIMARY_ROLLOUT
+      - name: ANALYSIS
+        with:
+          duration: 30m
+          metrics:
+            - provider: my-prometheus
+              interval: 10m
+              expected:
+                max: 0.1
+              query: rate(cpu_usage_total{app="foo"}[10m])
+```
+
+We can configure it as usual. No additional configuration needed.
+
+Note that the `expected` field can be used only for this `THRESHOLD` strategy.
+
+**The `PREVIOUS` strategy:**
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  pipeline:
+    stages:
+      - name: K8S_PRIMARY_ROLLOUT
+      - name: ANALYSIS
+        with:
+          duration: 30m
+          metrics:
+            - strategy: PREVIOUS
+              provider: my-prometheus
+              interval: 10m
+              deviation: HIGH
+              query: rate(cpu_usage_total{app="foo"}[10m])
+```
+
+**The `CANARY_BASELINE` strategy:**
+
+`CANARY_PRIMARY` will look almost the same.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  pipeline:
+    stages:
+      - name: K8S_CANARY_ROLLOUT
+        with:
+          replicas: 20%
+      - name: K8S_BASELINE_ROLLOUT
+        with:
+          replicas: 20%
+      - name: ANALYSIS
+        with:
+          duration: 30m
+          metrics:
+            - strategy: CANARY_BASELINE
+              provider: my-prometheus
+              interval: 10m
+              deviation: HIGH
+              query: rate(cpu_usage_total{app="foo", variant="{{ .Variant.Name }}"}[10m])
+      - name: K8S_PRIMARY_ROLLOUT
+      - name: K8S_CANARY_CLEAN
+      - name: K8S_BASELINE_CLEAN
+```
+
+**The pattern of using custom args:**
+
+You can reffer the defined custom args by using `{{ .Variant.CustomArgs.xxx }}`.
+
+```yaml
+      - name: ANALYSIS
+        with:
+          duration: 30m
+          metrics:
+            - strategy: CANARY_BASELINE
+              provider: my-prometheus
+              interval: 10m
+              deviation: HIGH
+              query: rate(cpu_usage_total{app="foo", bar="{{ .Variant.CustomArgs.bar }}"}[10m])
+              baselineArgs:
+                bar: 1
+              canaryArgs: 
+                bar: 2
+```
+
+**The pattern of using custom args in AnalysisTemplate:**
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: AnalysisTemplate
+spec:
+  metrics:
+    http_error_rate:
+      strategy: CANARY_BASELINE
+      provider: my-prometheus
+      interval: 10m
+      deviation: HIGH
+      query: |
+        sum without(status) (rate(http_requests_total{status=~"5.*", job="{{ .App.Name }}", bar="{{ .Variant.CustomArgs.bar }}"}[10m]))
+        /
+        sum without(status) (rate(http_requests_total{job="{{ .App.Name }}, bar="{{ .Variant.CustomArgs.bar }}"}[10m]))
```

Comment:
TBH, while reading this query it took me time to understand where the value will be defined for each argument.
For example, the value of `.App.Name`, `.Variant...`  is defined inside this template's Args or in the built-in Args or template caller's Args.

So how about applying an explicit name for each argument location to know where they are from?
We can discuss to find good names for them but at this point, they could be:

- `BuiltInArgs` or `GeneratedArgs` for built-in data, `.BultinArgs.Variant` to access variant name generated by Piped.
- `VariantArgs` for custom data defined by the user for each variant, e.g. `.VariantArgs.bar`.
- `TemplateArgs` for custom data defined by the user to use the template.

Maybe we should update the argument field name to unify with the above names too.
For examples:


``` yaml
      - name: ANALYSIS
        with:
          duration: 30m
          metrics:
            - strategy: CANARY_BASELINE
              provider: my-prometheus
              interval: 10m
              deviation: HIGH
              query: rate(cpu_usage_total{app="foo", bar="{{ .variantArgs.bar }}"}[10m])
              variantArgs:
                bar:
                  primary: 1
                  canary: 2
```

instead of

``` yaml
      - name: ANALYSIS
        with:
          duration: 30m
          metrics:
            - strategy: CANARY_BASELINE
              provider: my-prometheus
              interval: 10m
              deviation: HIGH
              query: rate(cpu_usage_total{app="foo", bar="{{ .Args.bar }}"}[10m])
              baselineArgs:
                bar: 1
              canaryArgs: 
                bar: 2
```

and 

``` yaml
     - template: 
          name: http_error_rate
          templateArgs:
            namespace: default
```

instead of 

``` yaml
     - template: 
          name: http_error_rate
          args:
            namespace: default
```

Reply:
Thank you! Looks cool. In the case of making a breaking change on AnalysisTemplate, let me re-think that names based on your idea.

Reply:
Sure. 👍

Reply:
I settled on applying `AppArgs` instead of `TemplateArg` because that is a group of variables where are different among Applications.
- `{{ .BuiltInArgs }}`
- `{{ .VariantArgs }}`
- `{{ .AppArgs }}`

Reply:
Looks good. 👍